Add journals pubmed api

wsf246 opened the pull request:
closes #53

Commit:
create publishers model

Commit:
add publisher_id to paper

Commit:
setup relationship between publisher and paper

Commit:
rename publisher to publication

Commit:
take out unneccesary uuid generation.

Commit:
use existing publication if exists

Commit:
force downcase on publication name

Commit:
add pubication to index show papers

Commit:
add source to pubmed api get

Commit:
capitalize journal name in index

briankung commented on app/controllers/papers_controller.rb:
```diff
@@ -18,6 +18,7 @@ def new
     @paper = Paper.new
     # SHAME SHAME SHAME DINGDINGDING SHAME SHAME SHAME DINGDINGDING
     5.times { @paper.authors.build }
+    @paper.build_publication
```
Let's not do this. We should do this in the view, if anything.

briankung replied:
Same for the line above this.

briankung commented on app/models/publication.rb:
```diff
@@ -0,0 +1,9 @@
+    before_save :downcase_fields
+
+    def downcase_fields
+        self.name.downcase!
```
Watch indentation here

briankung replied:
Also, <pedantic>this method name isn't exactly accurate, as it just downcases one field, right?</pedantic>

briankung commented on db/migrate/20160820205817_create_publishers.rb:
```diff
@@ -0,0 +1,10 @@
+      t.timestamps
+    end
+  end
+end
```
Are they publishers or publications?

wsf246 replied:
I tried that but the migration file name reverted back on the migrate, unless I did something funky.  I think it's some rails magic that ensures you don't fuck with migrations.

wsf246 commented on db/migrate/20160820205817_create_publishers.rb:
```diff
@@ -0,0 +1,10 @@
+      t.timestamps
+    end
+  end
+end
```
It was originally publishers, but then I rolledbacked and renamed it
publication.  Couldn't rename the migration and didn't want to create a new
one.  Let me know if there's a better way to this.

On Mon, Aug 22, 2016 at 10:28 PM, Brian Kung notifications@github.com
wrote:

> In db/migrate/20160820205817_create_publishers.rb
> https://github.com/briankung/projectcredo/pull/64#discussion_r75797832:
> 
> > @@ -0,0 +1,10 @@
> > +class CreatePublishers < ActiveRecord::Migration[5.0]
> > -  def change
> > -    create_table :publications, id: :uuid  do |t|
> > -      t.string :name
> > -      t.string :type
> >   +
> > -      t.timestamps
> > -    end
> > -  end
> >   +end
> 
> Are they publishers or publications?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/64/files/2b8cccb48a7d372bd7c56a40f6732f85f108ff4b#r75797832,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACKsZG9OrPGRh6Jj1Kdy4iip-ooAEuVhks5qimjqgaJpZM4JqR56
> .

briankung replied:
Why couldn't you rename the migration? Just roll back and then change the
filename and the migration class name.

On Tue, Aug 23, 2016 at 12:26 AM, wsf246 notifications@github.com wrote:

> In db/migrate/20160820205817_create_publishers.rb
> https://github.com/briankung/projectcredo/pull/64#discussion_r75800848:
> 
> > @@ -0,0 +1,10 @@
> > +class CreatePublishers < ActiveRecord::Migration[5.0]
> > -  def change
> > -    create_table :publications, id: :uuid  do |t|
> > -      t.string :name
> > -      t.string :type
> >   +
> > -      t.timestamps
> > -    end
> > -  end
> >   +end
> 
> It was originally publishers, but then I rolledbacked and renamed it
> publication. Couldn't rename the migration and didn't want to create a new
> one. Let me know if there's a better way to this.
> … <#m_2861241548378740647_>
> On Mon, Aug 22, 2016 at 10:28 PM, Brian Kung **_@**_.***> wrote: In
> db/migrate/20160820205817_create_publishers.rb <#64 (comment)
> https://github.com/briankung/projectcredo/pull/64#discussion_r75797832>:
> 
> > @@ -0,0 +1,10 @@ > +class CreatePublishers < ActiveRecord::Migration[5.0]
> > - def change > + create_table :publications, id: :uuid do |t| > +
> >   t.string :name > + t.string :type > + > + t.timestamps > + end > + end >
> >   +end Are they publishers or publications? — You are receiving this because
> >   you authored the thread. Reply to this email directly, view it on GitHub <
> >   https://github.com/briankung/projectcredo/pull/64/files/
> >   2b8cccb48a7d372bd7c56a40f6732f85f108ff4b#r75797832>, or mute the thread <
> >   https://github.com/notifications/unsubscribe-
> >   auth/ACKsZG9OrPGRh6Jj1Kdy4iip-ooAEuVhks5qimjqgaJpZM4JqR56> .
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/64/files/2b8cccb48a7d372bd7c56a40f6732f85f108ff4b#r75800848,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACtGx-wxBN0bFymyLTgcnNLg-0Tm7x60ks5qinZzgaJpZM4JqR56
> .

## 

Brian Kung
http://about.me/briankung/
http://www.facebook.com/brian.kung http://www.linkedin.com/in/briankung
http://www.twitter.com/briankung

BE A UNICORN!

briankung replied:
That doesn't seem right. Are you sure it wasn't a git thing?

I'm 99% sure Rails doesn't do that.

On Tue, Aug 23, 2016 at 10:51 AM, wsf246 notifications@github.com wrote:

> In db/migrate/20160820205817_create_publishers.rb
> https://github.com/briankung/projectcredo/pull/64#discussion_r75882389:
> 
> > @@ -0,0 +1,10 @@
> > +class CreatePublishers < ActiveRecord::Migration[5.0]
> > -  def change
> > -    create_table :publications, id: :uuid  do |t|
> > -      t.string :name
> > -      t.string :type
> >   +
> > -      t.timestamps
> > -    end
> > -  end
> >   +end
> 
> I tried that but the migration file name reverted back on the migrate,
> unless I did something funky. I think it's some rails magic that ensures
> you don't fuck with migrations.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/64/files/2b8cccb48a7d372bd7c56a40f6732f85f108ff4b#r75882389,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACtGx1WpixBULXC_mN6FkRfi1CbElyV0ks5qiwjegaJpZM4JqR56
> .

## 

Brian Kung
http://about.me/briankung/
http://www.facebook.com/brian.kung http://www.linkedin.com/in/briankung
http://www.twitter.com/briankung

BE A UNICORN!

wsf246 replied:
possibly a git thing.  let me try it again and see what happens.

On Tue, Aug 23, 2016 at 9:53 AM, Brian Kung notifications@github.com
wrote:

> In db/migrate/20160820205817_create_publishers.rb
> https://github.com/briankung/projectcredo/pull/64#discussion_r75882861:
> 
> > @@ -0,0 +1,10 @@
> > +class CreatePublishers < ActiveRecord::Migration[5.0]
> > -  def change
> > -    create_table :publications, id: :uuid  do |t|
> > -      t.string :name
> > -      t.string :type
> >   +
> > -      t.timestamps
> > -    end
> > -  end
> >   +end
> 
> That doesn't seem right. Are you sure it wasn't a git thing? I'm 99% sure
> Rails doesn't do that.
> … <#m_7746067793841332578_>
> On Tue, Aug 23, 2016 at 10:51 AM, wsf246 **_@**_.***> wrote: In
> db/migrate/20160820205817_create_publishers.rb <#64 (comment)
> https://github.com/briankung/projectcredo/pull/64#discussion_r75882389>:
> 
> > @@ -0,0 +1,10 @@ > +class CreatePublishers < ActiveRecord::Migration[5.0]
> > - def change > + create_table :publications, id: :uuid do |t| > +
> >   t.string :name > + t.string :type > + > + t.timestamps > + end > + end >
> >   +end I tried that but the migration file name reverted back on the migrate,
> >   unless I did something funky. I think it's some rails magic that ensures
> >   you don't fuck with migrations. — You are receiving this because you
> >   commented. Reply to this email directly, view it on GitHub <
> >   https://github.com/briankung/projectcredo/pull/64/files/
> >   2b8cccb48a7d372bd7c56a40f6732f85f108ff4b#r75882389>, or mute the thread <
> >   https://github.com/notifications/unsubscribe-auth/ACtGx1WpixBULXC_
> >   mN6FkRfi1CbElyV0ks5qiwjegaJpZM4JqR56> .
> >   -- Brian Kung http://about.me/briankung/ <http://www.facebook.com/
> >   brian.kung> http://www.linkedin.com/in/briankung <
> >   http://www.twitter.com/briankung> BE A UNICORN!
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/64/files/2b8cccb48a7d372bd7c56a40f6732f85f108ff4b#r75882861,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACKsZHxOwd4tPQon081LzQrdRXFt8yKzks5qiwlZgaJpZM4JqR56
> .

briankung commented on app/controllers/references_controller.rb:
```diff
@@ -41,7 +41,7 @@ def import_paper identifier
       data = pubmed.search(identifier)['result'][identifier]
 
-      existing_authors, new_authors = [], []
+      existing_authors, new_authors, publication = [], [], []
```
Do you do anything with this array?

wsf246 replied:
nooope, vestige of something I was attempting.  removed.

briankung commented on app/models/paper.rb:
```diff
@@ -43,4 +45,8 @@ def valid_methodologies
   end
+
+  def find_publication
+    self.publication = Publication.where(name: self.publication.name).first_or_initialize
```
What happens if `self.publication` is `nil` and you call `self.publication.name`?

wsf246 replied:
it breaks, fixing it now.

wsf246 replied:
yea, before validation.  This will let it select an existing publication instead of creating a new one.

briankung commented on app/views/papers/_form.html.erb:
```diff
@@ -23,6 +23,13 @@
     </div>
   <% end %>
 
+  <%= f.fields_for :publication, paper.publication ? paper.publication : paper.build_publication do |p| %>
```
You could also do `(paper.publication || paper.build_publication)`

briankung commented on app/views/papers/_index.html.erb:
```diff
@@ -32,8 +32,8 @@
         <i>
-          in Journal of Reserved
-          (<%= paper.published_at %>)
+          <%= "in " + paper.publication.name.split.map(&:capitalize).join(' ') unless paper.publication.nil? %>
```
Let's do this in CSS instead https://developer.mozilla.org/en-US/docs/Web/CSS/text-transform#Values

briankung commented on app/views/papers/_index.html.erb:
```diff
@@ -32,8 +32,8 @@
-          in Journal of Reserved
-          (<%= paper.published_at %>)
+          <%= "in " + paper.publication.name.split.map(&:capitalize).join(' ') unless paper.publication.nil? %>
+          <%= "(" + paper.published_at.to_s + ")" unless paper.published_at.nil? %>
```
Should this ever be `nil`?

wsf246 replied:
yea, sometimes pubmed doesn't have it, so unless we add a default, they will be nil.

briankung commented on app/views/papers/show.html.erb:
```diff
@@ -9,6 +9,11 @@
 
 <p>
+  <strong>Publication:</strong>
+  <%= @paper.publication.name.split.map(&:capitalize).join(' ') %>
```
Ditto here https://developer.mozilla.org/en-US/docs/Web/CSS/text-transform#Values

briankung commented on test/fixtures/publishers.yml:
```diff
@@ -0,0 +1,9 @@
+
+two:
+  name: MyString
+  type: 
```
My policy recently with these generated files is to discard them if we're not actively using them.

Commit:
remove extra publication array

Commit:
move builds out of paper controller to view

Commit:
rename downcase_fields to downcase_names

briankung commented on app/models/paper.rb:
```diff
@@ -43,4 +45,8 @@ def valid_methodologies
   end
+
+  def find_publication
+    self.publication = Publication.where(name: self.publication.name).first_or_initialize
```
Before you fix it, do you use this anywhere?

On Tue, Aug 23, 2016, 12:12 PM wsf246 notifications@github.com wrote:

> In app/models/paper.rb
> https://github.com/briankung/projectcredo/pull/64#discussion_r75899350:
> 
> > @@ -43,4 +45,8 @@ def valid_methodologies
> >        'case report'
> >      ]
> >    end
> > +
> > -  def find_publication
> > -    self.publication = Publication.where(name: self.publication.name).first_or_initialize
> 
> it breaks, fixing it now.
> 
> —
> You are receiving this because you commented.
> 
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/64/files/2b8cccb48a7d372bd7c56a40f6732f85f108ff4b#r75899350,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACtGx5YVRK3gIEawX89XTkiZ4b9PDwnRks5qixvcgaJpZM4JqR56
> .

Commit:
find publication only when publication name is passed in by params

Commit:
put capitalize publication in css

Commit:
rename publisher migrations to publication

Commit:
remove unused files

wsf246 commented:
That should be everything.  Let me know what you think!

briankung commented on app/models/paper.rb:
```diff
@@ -43,4 +45,10 @@ def valid_methodologies
+
+  def find_publication
+    if self.publication.present?
+      self.publication = Publication.where(name: self.publication.name).first_or_initialize
```
I'm still not entirely sure why you're doing this, but I think you can do this: http://guides.rubyonrails.org/active_record_querying.html#find-or-initialize-by

briankung replied:
This whole `before_validation` hook seems very wrong. Why is this necessary?

briankung commented on app/models/paper.rb:
```diff
@@ -5,10 +5,12 @@ class Paper < ApplicationRecord
 
   accepts_nested_attributes_for :authors, reject_if: proc { |attributes| attributes['name'].blank? }
-  validates_associated :authors
+  accepts_nested_attributes_for :publication, reject_if: proc { |attributes| attributes['name'].blank? }
```
You use `proc { |attributes| attributes['name'].blank? }` in two places. Should we think about maybe assigning the proc to a variable or just encapsulating it as a method?

wsf246 replied:
yea, we can do that.

On Tue, Aug 23, 2016 at 2:55 PM, Brian Kung notifications@github.com
wrote:

> In app/models/paper.rb
> https://github.com/briankung/projectcredo/pull/64#discussion_r75938547:
> 
> >    accepts_nested_attributes_for :authors, reject_if: proc { |attributes| attributes['name'].blank? }
> > -  validates_associated :authors
> > -  accepts_nested_attributes_for :publication, reject_if: proc { |attributes| attributes['name'].blank? }
> 
> You use proc { |attributes| attributes['name'].blank? } in two places.
> Should we think about maybe assigning the proc to a variable or just
> encapsulating it as a method?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/64/files/c10ff7dd636d241cfd3c03b300def8df2400222c#r75938547,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACKsZOiX4qNRv4zaHBxAVV20WSuiAVd5ks5qi1AegaJpZM4JqR56
> .

wsf246 commented on app/models/paper.rb:
```diff
@@ -43,4 +45,10 @@ def valid_methodologies
+
+  def find_publication
+    if self.publication.present?
+      self.publication = Publication.where(name: self.publication.name).first_or_initialize
```
So I kinda just did stuff until I got the behavior I was expecting.  I
originally started with a validate_uniqueness in publications.  Problem was
if I tried to save a paper with an existing publication, it would throw an
error.  So I added this before_validation to get it to get the params info
before the validation because the validation would throw away the data
before it got to the controller.  That was still causing an error as the
nested attribute wouldn't let me save with a different id when I was
updating.

So I got rid of the validate_uniqueness and it all seemed to work.  I
figure this method would keep us from creating the same journal multiple
times.  It should I think also work for the Pubmed API because this method
is sitting in the model?  Anyway, we can partner code this as well cuz I
took multiple rounds to try to get this functionality to work but I'm only
60% conscious of all the ramifications.

On Tue, Aug 23, 2016 at 2:51 PM, Brian Kung notifications@github.com
wrote:

> In app/models/paper.rb
> https://github.com/briankung/projectcredo/pull/64#discussion_r75937890:
> 
> > @@ -43,4 +45,10 @@ def valid_methodologies
> >        'case report'
> >      ]
> >    end
> > +
> > -  def find_publication
> > -    if self.publication.present?
> > -      self.publication = Publication.where(name: self.publication.name).first_or_initialize
> 
> I'm still not entirely sure why you're doing this, but I think you can do
> this: http://guides.rubyonrails.org/active_record_querying.html#
> find-or-initialize-by
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/64/files/c10ff7dd636d241cfd3c03b300def8df2400222c#r75937890,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACKsZLmMj7szfS2wkSN3lIWH0LLcqOanks5qi08qgaJpZM4JqR56
> .

wsf246 replied:
Once I took out the validate_uniqueness from publication.rb, I think we
prob don't need to do it before_validation?  Where's the best place to put
it if we don't need it before validation?

On Tue, Aug 23, 2016 at 2:58 PM, Brian Kung notifications@github.com
wrote:

> In app/models/paper.rb
> https://github.com/briankung/projectcredo/pull/64#discussion_r75939090:
> 
> > @@ -43,4 +45,10 @@ def valid_methodologies
> >        'case report'
> >      ]
> >    end
> > +
> > -  def find_publication
> > -    if self.publication.present?
> > -      self.publication = Publication.where(name: self.publication.name).first_or_initialize
> 
> This whole before_validation hook seems very wrong. Why is this necessary?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/64/files/c10ff7dd636d241cfd3c03b300def8df2400222c#r75939090,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACKsZBm91QW9AoxQqMK5kQw9BN6a8TSBks5qi1DhgaJpZM4JqR56
> .

briankung commented on app/views/papers/_form.html.erb:
```diff
@@ -16,13 +16,20 @@
   </div>
 
-  <%= f.fields_for :authors, paper.authors do |a| %>
+  <%= f.fields_for :authors, (paper.authors.empty? ? 5.times { paper.authors.build } : paper.authors) do |a| %>
```
We probably want to show `no. of authors + 1`, right?

Also, `5.times { paper.authors.build }` evaluates to `#=> 5`, and I'm not sure how `fields_for` handles an integer as an input.

Commit:
uses autosave functionality

wsf246 commented:
so this seems to do what I want it to do, a few things though:

1) I haven't found a good way for it to reject if blank outside of the "reject_if" method.  I'm not sure how to call a silent rejection in the "autosave_associated_records_for" method, if I don't use the reject_if, it will throw an exception "name can't be blank".

2) I'm still not sure what the autosave: true/false does.  I looked through the documentation and tested it without seeing any changes in functionality, would love your opinion.

3) I changed the autosave to create a new record for any new name.  Before if I updated "brian's journal" to "brian journal" if it wasn't an existing record, it would save over the existing record with the new name instead of creating a new one.  Now it creates a new record.

4) The functionality appears to work from everything I wanted it to do.  I think in the future, I should have done what you had suggested and just have made publication a string in paper.  But for now, this seems to work and I'm probably 85% for using this branch and 15% using just a paper attribute for simplicity.  If it turns out this is causing more headaches down the road, we can switch to just using a paper attribute.

Thoughts?

briankung commented on app/models/paper.rb:
```diff
@@ -43,4 +44,16 @@ def valid_methodologies
+      new_publication.name = publication.name
+      new_publication.save!
+      self.publication = new_publication
+    end
```
You can do the last two conditions all in one line, too:

``` ruby
self.publication = Publication.find_or_create_by name: publication.name.downcase
```

briankung replied:
And you can wrap that line in a `if publication` to catch the `nil` case.

briankung commented:
1. You have taken over the responsibilities of `reject_if` by overwriting `autosave_associated_records_for_publication`. That method should now be responsible for handling that.
2. Autosave has more to do with the parent object, in this case the `Publication`. Basically means that if you save the publication it will save the corresponding `Papers`. This doesn't actually have anything to do with `autosave_associated_records_for_publication`. 
3. OK
4. This should work but we should have some process to discuss scope ahead of time. Maybe have a pre-coding letter of intent on the GitHub issue for what we're planning on doing.

briankung commented:
I didn't get a chance to test this locally since I reformatted my Mac. Take feedback with a grain of salt. Test, test, test

Commit:
refactor autosave for associated nested attributes

briankung commented:
Closing in favor of #67

briankung commented:
Leaving the branch for future reference
